### PR TITLE
Re-add Firefly support by updating the api for the new Firefly-vis pypi project

### PR DESCRIPTION
**Pull request by agurvich**
Replaced references to `firefly_api` with references to the new `Firefly` (or `Firefly-vis`, as it's known on PyPi) package which contains source files for Firefly and an extended Python API for launching Flask webservers.

## PR Summary

`firefly_api` has been deprecated, so much so I tried to remove all traces of its existence from the internet.
That broke the yt CI. My bad. Now that I have uploaded the Firefly source files and the new api to PyPi
I needed to update how yt integrates with Firefly.

should replace: https://github.com/yt-project/yt/pull/3419

## PR Checklist

- [x] New features are documented, with docstrings and narrative docs
- [x] Adds a test for any bugs fixed. Adds tests for new features.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
Thanks @agurvich ! 
Beware you have a few linting errors listed here https://results.pre-commit.ci/run/github/88776493/1626285552.mncOcy2kTYa6qE-DfXDYjA
almost all of which can be solved using the comment you'll find at the bottom of the page, with the exception of the single flake8 error

**Comment by agurvich** [comment]
pre-commit.ci autofix

**Comment by agurvich** [comment]
@neutrinoceros Will that comment fix it?

**Comment by agurvich** [comment]
Running flake8 comes up with tons of errors in files that I haven't touched... A little confused how to fix that part
image

**Comment by neutrinoceros** [comment]
You shouldn't run flake8 directly. We use it via pre-commit. See https://yt-project.org/doc/developing/developing.html#automatically-checking-and-fixing-code-style

**Comment by neutrinoceros** [comment]
Anyway, the pre-commit bot is now happy so I'll review this now !

File: doc/source/visualizing/visualizing_particle_datasets_with_firefly.rst [hunk 1]
```diff
@@ -5,28 +5,22 @@ Visualizing Particle Datasets with Firefly
 `Firefly <https://github.com/ageller/Firefly>`_
 is an interactive, browser-based,
 particle visualization platform that allows you to filter, colormap, and fly
-through their data. The user interface can be entirely customized through the
-`python api <https://github.com/agurvich/firefly_api>`_
+through their data. The python frontend allows users to both load in their
```
Comment (neutrinoceros) [inline doc/source/visualizing/visualizing_particle_datasets_with_firefly.rst:8]:
```suggestion
through their data. The Python frontend allows users to both load in their
```

File: doc/source/visualizing/visualizing_particle_datasets_with_firefly.rst [hunk 2]
```diff
@@ -5,28 +5,22 @@ Visualizing Particle Datasets with Firefly
...
 to export your data to Firefly's JSON format through the
 :meth:`~yt.data_objects.data_containers.YTDataContainer.create_firefly_object`
-method. You will still need to clone the
-`repository <https://github.com/ageller/Firefly>`_
-and ensure these data files are stored in the ``Firefly/data`` directory,
-this is easily done using the
-``path_to_firefly`` argument of
-:meth:`~yt.data_objects.data_containers.YTDataContainer.create_firefly_object`
-which should point at Firefly's ``index.html``.
+method.
```
Comment (neutrinoceros) [inline doc/source/visualizing/visualizing_particle_datasets_with_firefly.rst:13]: love it

File: yt/data_objects/tests/test_firefly.py [hunk 3]
```diff
@@ -3,16 +3,38 @@
...
+
+    reader.dumpToJSON(write_jsons_to_disk=False)
+
+    if len(reader.JSON) == 0:
+        raise ValueError(
+            "reader.JSON was not output to string correctly."
+            + " either Firefly is damaged or needs a hotfix-- try reinstalling."
+            + " if that doesn't work contact the developers"
+            + " at github.com/ageller/Firefly/issues."
+        )
```
Comment (neutrinoceros) [inline yt/data_objects/tests/test_firefly.py:25]: I think this should be split into an assert statement and a comment

**Review by neutrinoceros — CHANGES_REQUESTED** [review]
Good job !  Besides a couple nitpicks, I think this is the longterm solution we needed, thank you !

File: yt/data_objects/tests/test_firefly.py (hunk 3 above)
Comment (agurvich) [reply yt/data_objects/tests/test_firefly.py]: okay! Made the change :)

**Comment by agurvich** [comment]
pre-commit.ci autofix

**Review by neutrinoceros — APPROVED** [review]

**Comment by agurvich** [comment]
sorry i got trigger happy cleaning up my repo 👀

**Comment by neutrinoceros** [comment]
I was playing with the idea of merging this, but I think now that we would need a second reviewer since I don't think the code that's changed by the patch isn't really _broken_ at the moment, so I'm triaging this as a "refactor".

**Comment by agurvich** [comment]
If you'd like I can delete the firefly_api pypi package again 😛

**Comment by neutrinoceros** [comment]
lol. Please don't :)

File: setup.cfg [hunk 4]
```diff
@@ -71,6 +71,7 @@ doc =
     sphinx-bootstrap-theme
     sphinx-rtd-theme
 full =
+    Firefly-vis>=2.0.0
```
Comment (neutrinoceros) [inline setup.cfg:74]: @agurvich could you use lower case here and sort in in alpha order please ?
Sorry for the late nitpick
```suggestion
    firefly-vis>=2.0.0
```

**Review by neutrinoceros — APPROVED** [review]

File: yt/data_objects/data_containers.py [hunk 5]
```diff
@@ -772,35 +769,35 @@ def create_firefly_object(
...
         except ImportError as e:
             raise ImportError(
-                "Can't find firefly_api, ensure it "
+                "Can't find Firefly, ensure it "
                 "is in your python path or install it with "
-                "`python -m pip install firefly_api`. It is also available "
-                "on github at github.com/agurvich/firefly_api"
+                "`python -m pip install Firefly-vis`. It is also available "
+                "on github at github.com/ageller/Firefly"
             ) from e
```
Comment (cphyc) [inline yt/data_objects/data_containers.py:794]: I know this isn't firefly-specific and was already there, but we have a way of lazy-importing packages in yt (`yt/utilities/on_demand_imports.py`). This block of code should probably be moved there.

File: yt/data_objects/tests/test_firefly.py [hunk 6]
```diff
@@ -3,16 +3,36 @@
 from yt.testing import fake_particle_ds, requires_module
 
 
-@requires_module("firefly_api")
-def test_firefly_JSON_object():
+@requires_module("Firefly")
```
Comment (cphyc) [inline yt/data_objects/tests/test_firefly.py:6]: These tests require `Firefly` to run, but they won't run on the CI unless you add Firefly to the requirements in `setup.cfg`. I think adding Firefly as a dependency in 
```
[options.extras_require]
…
full =
    …
    firefly-vis>=MINIMAL_VERSION
```

Comment (neutrinoceros) [reply yt/data_objects/tests/test_firefly.py:6]: this is already part of the PR @cphyc

Comment (cphyc) [reply yt/data_objects/tests/test_firefly.py:6]: It is indeed. For some reason it didn't show up in my review? Or maybe I was stupid? We'll never know!

File: yt/data_objects/data_containers.py (hunk 5 above)
Comment (agurvich) [reply yt/data_objects/data_containers.py]: oh for sure, i did not realize that existed. much much cleaner now

**Comment by agurvich** [comment]
pre-commit.ci autofix

**Comment by agurvich** [comment]
question: can i avoid having to comment that autofix thing if i set a configuration somewhere or add it into my commit message? I tried googling but i couldn't find anything

File: yt/data_objects/data_containers.py [hunk 7]
```diff
@@ -782,16 +782,7 @@ def create_firefly_object(
...
-        try:
-            from Firefly.data_reader import ParticleGroup, Reader
-        except ImportError as e:
-            raise ImportError(
-                "Can't find Firefly, ensure it "
-                "is in your python path or install it with "
-                "`python -m pip install Firefly-vis`. It is also available "
-                "on github at github.com/ageller/Firefly"
-            ) from e
+        from yt.utilities.on_demand_imports import _firefly as firefly
```
Comment (neutrinoceros) [inline yt/data_objects/data_containers.py:785]: Thanks for patching this. Another remark is that the way we implement our "on_demand_imports" module, this import is perfectly safe at the top level of the module, it doesn't need to be embedded locally. Can you move it please ? (we're getting there :))

File: yt/utilities/on_demand_imports.py [hunk 8]
```diff
@@ -665,3 +665,26 @@ def __getattr__(self, attr):
 
 
 _pandas = pandas_imports()
+
+class firefly_imports:
```
Comment (neutrinoceros) [inline yt/utilities/on_demand_imports.py:670]: I think this class is missing a `__init__` method. Additionally, I don't think the implemented properties mocking submodules are necessary, see how the `pandas_import` class above for a minimalist working implementation.

**Comment by cphyc** [comment]
> question: can i avoid having to comment that autofix thing if i set a configuration somewhere or add it into my commit message? I tried googling but i couldn't find anything

@agurvich yes you can using pre-commit hooks, see https://yt-project.org/docs/dev/developing/developing.html#automatically-checking-and-fixing-code-style!

File: yt/utilities/on_demand_imports.py (hunk 8 above)
Comment (agurvich) [reply yt/utilities/on_demand_imports.py:670]:
```
class firefly_imports:
    _name = "firefly"
    _module = None

    def __init__(self):
        try:
            import Firefly as myself

            self._module = myself
        except ImportError:
            self._module = NotAModule(self._name)

    def __getattr__(self,attr):
        return getattr(self._module,attr)


_firefly = firefly_imports()
```

Does not work and raises an attribute error for the sub-module `data_reader`. I based my original implementation on the [yaml imports](https://github.com/yt-project/yt/blob/main/yt/utilities/on_demand_imports.py#L554) which does not raise an attribute error.

File: yt/data_objects/data_containers.py (hunk 7 above)
Comment (agurvich) [reply yt/data_objects/data_containers.py:785]: good call

**Review by cphyc — APPROVED** [review]

**Pull request merged**